### PR TITLE
Fix Software and Hardware Mismatch

### PR DESCRIPTION
**Pull request by definelicht**
Apparently the compiler didn't like the `PackedFloat` struct, so I shoehorned it into a different data type that packs everything into an `ap_uint`, and otherwise does everything exactly the same. This seems to fix it.

### CODE REVIEW

File: device/MatrixMultiplication.cpp [hunk 1]
```diff
@@ -283,7 +286,7 @@ void WriteCInner(hlslib::Stream<PackedFloat> &from_kernel, DramLine *const mem,
 #pragma HLS PIPELINE II = 1
 #pragma HLS LOOP_FLATTEN
             if (i == 0) {
-                *reinterpret_cast<PackedFloat *>(num) = from_kernel.Pop();
+                from_kernel.Pop() >> num;
```
Comment (ChrisPattison) [inline device/MatrixMultiplication.cpp]: I'm not sure this usage of the `>>` operator is consistent with the Xilinx Stream class. See pg 228 of https://www.xilinx.com/support/documentation/sw_manuals/xilinx2017_1/ug902-vivado-high-level-synthesis.pdf

File: host/UnitTests.cpp [hunk 2]
```diff
@@ -22,11 +22,11 @@ TEST_CASE("PackedFloat to/from GMP Conversion") {
         REQUIRE(gmp_num->_mp_size == -1);
         REQUIRE(gmp_num->_mp_d[0] == 42);
         PackedFloat num(gmp_num);
-        REQUIRE(num.sign == 1);
-        REQUIRE(num.exponent == 1);
-        REQUIRE(num.mantissa[0] == 42);
-        for (int i = 1; i < kMantissaBytes; ++i) {
-            REQUIRE(num.mantissa[i] == 0);
+        REQUIRE(num.GetSign() == 1);
```
Comment (ChrisPattison) [inline host/UnitTests.cpp:25]: I would prefer that we test for nonzero. This seems a bit brittle

Edit: I guess this was already here so `¯\_(ツ)_/¯`

File: include/Types.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#pragma once
```
Comment (ChrisPattison) [inline include/Types.h]: Can we name this `DeviceTypes.h` or something? `Types.h` is pretty ambiguous

File: device/MatrixMultiplication.cpp (hunk 1 above)
Comment (definelicht) [reply device/MatrixMultiplication.cpp]: This is `operator>>` from `PackedFloat`, it has nothing to do with the stream class, but I can make it explicit and call it `UnpackFlits` or something

File: include/Types.h (hunk 3 above)
Comment (definelicht) [reply include/Types.h]: Sure

File: device/MatrixMultiplication.cpp (hunk 1 above)
Comment (ChrisPattison) [reply device/MatrixMultiplication.cpp]: Ah, yeah. Probably better to just make it explicit.

**Pull request merged**